models/issue_dependency.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2018 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/util"
+)
+
+// IssueDependency represents an issue dependency
+type IssueDependency struct {
+	ID           int64          `xorm:"pk autoincr"`
+	UserID       int64          `xorm:"NOT NULL"`
+	IssueID      int64          `xorm:"UNIQUE(issue_dependency) NOT NULL"`
+	DependencyID int64          `xorm:"UNIQUE(issue_dependency) NOT NULL"`
+	CreatedUnix  util.TimeStamp `xorm:"INDEX created"`
+	UpdatedUnix  util.TimeStamp `xorm:"updated"`
+}
+
+// DependencyType Defines Dependency Type Constants
+type DependencyType int
+
+// Define Dependency Types
+const (
+	DependencyTypeBlockedBy DependencyType = iota
+	DependencyTypeBlocking
+)
+
+// CreateIssueDependency creates a new dependency for an issue
+func CreateIssueDependency(user *User, issue, dep *Issue) (err error) {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	// Check if it aleready exists
+	err = issueDepExists(sess, issue.ID, dep.ID)
+	if err != nil {
+		return
+	}
+
+	// No error means the dependency doesn't exist nor is circular, so we can proceed to create a new one.
+	newIssueDependency := &IssueDependency{
+		UserID:       user.ID,
+		IssueID:      issue.ID,
+		DependencyID: dep.ID,
+	}
+
+	if _, err := sess.Insert(newIssueDependency); err != nil {
+		return err
+	}
+
+	// Add comment referencing the new dependency
+	if _, err = createIssueDependencyComment(sess, user, issue, dep, true); err != nil {
+		return
+	}
+
+	// Create a new comment for the dependent issue
+	if _, err = createIssueDependencyComment(sess, user, dep, issue, true); err != nil {
+		return
+	}
+
+	return sess.Commit()
+}
+
+// RemoveIssueDependency removes a dependency from an issue
+func RemoveIssueDependency(user *User, issue *Issue, dep *Issue, depType DependencyType) (err error) {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	// Check if it exists
+	var exists bool
+	switch depType {
+	case DependencyTypeBlockedBy:
+		err = issueDepExists(sess, issue.ID, dep.ID)
+	case DependencyTypeBlocking:
+		err = issueDepExists(sess, dep.ID, issue.ID)
+	default:
+		return
```

Comment:
Missing error?

Reply:
Yes, kind of. The will never happen a case where this returns `nil` where it should be an error (the way the function is implemented), there is already a check [here](https://github.com/kolaente/gitea/blob/master/routers/repo/issue_dependency.go#L95).
But I've added a new custom error type which will be returned in the case you pointed out. Thanks for mentioning that.